core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableMergeUnion.java

```diff
@@ -0,0 +1,128 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.adapter.enumerable;
+
+import org.apache.calcite.linq4j.Ord;
+import org.apache.calcite.linq4j.tree.BlockBuilder;
+import org.apache.calcite.linq4j.tree.Expression;
+import org.apache.calcite.linq4j.tree.Expressions;
+import org.apache.calcite.linq4j.tree.ParameterExpression;
+import org.apache.calcite.plan.RelOptCluster;
+import org.apache.calcite.plan.RelOptCost;
+import org.apache.calcite.plan.RelOptPlanner;
+import org.apache.calcite.plan.RelTraitSet;
+import org.apache.calcite.rel.RelCollation;
+import org.apache.calcite.rel.RelCollationTraitDef;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.metadata.RelMetadataQuery;
+import org.apache.calcite.util.BuiltInMethod;
+import org.apache.calcite.util.Pair;
+import org.apache.calcite.util.Util;
+
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/** Implementation of {@link org.apache.calcite.rel.core.Union} in
+ * {@link org.apache.calcite.adapter.enumerable.EnumerableConvention enumerable calling convention}.
+ * Performs a union (or union all) of all its inputs (which must be already sorted),
+ * respecting the order. */
+public class EnumerableMergeUnion extends EnumerableUnion {
+
+  protected EnumerableMergeUnion(RelOptCluster cluster, RelTraitSet traitSet, List<RelNode> inputs,
+      boolean all) {
+    super(cluster, traitSet, inputs, all);
+    final List<RelCollation> collations = traitSet.getTraits(RelCollationTraitDef.INSTANCE);
+    if (collations == null || collations.isEmpty()) {
+      throw new IllegalArgumentException("EnumerableMergeUnion with no collation");
+    }
+  }
+
+  public static EnumerableMergeUnion create(RelCollation collation, List<RelNode> inputs,
+      boolean all) {
+    final RelOptCluster cluster = inputs.get(0).getCluster();
+    final RelTraitSet traitSet = cluster.traitSetOf(EnumerableConvention.INSTANCE).replace(
+        collation);
+    return new EnumerableMergeUnion(cluster, traitSet, inputs, all);
+  }
+
+  @Override public @Nullable RelOptCost computeSelfCost(RelOptPlanner planner,
+      RelMetadataQuery mq) {
+    // EnumerableUnion computes the whole union result in memory, whereas EnumerableMergeUnion
+    // does it step by step, loading only 1 item per input at a time, reflect this optimization
+    // on the cpu cost:
+    final RelOptCost cost = super.computeSelfCost(planner, mq);
+    if (cost == null) {
+      return null;
+    }
+    return planner.getCostFactory().makeCost(cost.getRows(), cost.getCpu() / 2, cost.getIo());
```

Comment:
I believe this `/2` should not be applied for `all=true` case.

Reply:
So the only difference in computation complexity appears when `all=false`.
`EnumerableUnion` needs to keep the full lookup table in memory (it needs to perform `hashSet.add` for every row), while `EnumerableMergeUnion` needs to track the previously emitted row and emit a new row in case the next one is different. `MergeUnion` still has to call `equals`, so the thing does not sound like `two times cheaper`.

It looks like the `self` costs of both variations are more or less the same.

Reply:
Ok, I'll remove this overridden method. Let EnumerableMergeUnion have by default the same cost as EnumerableUnion.

Reply:
Method removed.